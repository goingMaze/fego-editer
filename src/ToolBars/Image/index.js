import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { AtomicBlockUtils } from 'draft-js'
import { ToolBarBtn } from '../../utils'
import Iconfont from '../../iconfont'

const IMAGE_TYPES = [
	{ label: 'image', style: 'image', icon: 'iconImage', title: '图片' }
], defaultState = {
	imgVisible: false,
	urlValue: '',
	imgUrl: '点击上传图片',
	btnDisabled: true,
	imgAddType: 'local',
	width: 'auto',
	height: 'auto'
}

export default class Image extends Component {
	static propTypes = {
		editorState: PropTypes.object.isRequired,
		onChange: PropTypes.func.isRequired,
		config: PropTypes.object.isRequired,
		modalManage: PropTypes.object.isRequired
	}
	constructor(props) {
		super(props)
		this.state = defaultState
	}
	componentWillMount() {
		let { modalManage } = this.props;
		modalManage.addCallback(this.changeImgVisible)
	}
	componentWillUnmount() {
		let { modalManage } = this.props;
		modalManage.removeCallback(this.changeImgVisible)
	}
	changeImgVisible = () => {
		this.setState({
			...defaultState,
			imgVisible: this.imgVisible
		})
		this.imgVisible = false
	}
	confirmMedia = () => {
		const { editorState, onChange } = this.props;
		const { imgUrl, urlType, width, height } = this.state;
		const contentState = editorState.getCurrentContent();
		const entityKey = contentState.createEntity(urlType, 'IMMUTABLE', { src: imgUrl, width, height }).getLastCreatedEntityKey();
		onChange(AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' '))
		this.changeImgVisible()
	}
	handleImgAdd = () => {
		let { uploadUrl = '', uploadCallBack = null } = this.props.config;
		let { imgAddType } = this.state;
		if (imgAddType === 'local' && uploadUrl) {
			let callback = imgUrl => this.setState({ imgUrl }, this.confirmMedia);
			uploadCallBack ? uploadCallBack(uploadUrl, this.refs.url, callback) :
				this.uploadCallBack(uploadUrl, this.refs.url, callback)
		} else {
			this.confirmMedia()
		}
	}
	uploadCallBack = (url, fileName, callback) => {
		const xhr = new XMLHttpRequest(), data = new FormData();
		xhr.open('POST', url)
		data.append('upload', fileName.file)
		xhr.addEventListener('load', () => {
			console.log('上传成功', 2)
			const response = JSON.parse(xhr.responseText);
			let { imgUrl } = response.data
			callback(imgUrl)
		})
		xhr.addEventListener('error', () => {
			const error = JSON.parse(xhr.responseText);
			console.log(error);
		})
		xhr.send(data)
	}
	handleChange = (type, e) => {
		let { value } = e.target
		if (type === 'imgUrl') {
			this.setState({
				urlValue: value,
				btnDisabled: !Boolean(value)
			})
		}
		this.setState({ [type]: value })
	}
	onToggle = () => {
		this.imgVisible = !this.state.imgVisible
		this.setState({
			urlType: 'image',
			urlValue: ''
		})
	}
	toggleImgAddType = e => {
		this.setState({
			imgAddType: e.target.value,
			imgUrl: '点击上传图片'
		})
		return false
	}
	render() {
		let { imgVisible, urlValue, btnDisabled, imgAddType, imgUrl, width, height } = this.state;
		let { accept = 'image/png, image/jpeg, image/jpg' } = this.props;
		let imageModal = '';
		if (imgVisible) {
			imageModal = (
				<div className='modal' ref='imgModal' onMouseDown={e => e.stopPropagation()} title='图片' >
					<div className='center' onMouseDown={this.toggleImgAddType} >
						<button className='btn' value='local' >本地图片</button>&nbsp;&nbsp;|&nbsp;&nbsp;
						<button className='btn' value='online' >在线图片</button>
					</div>
					<div className='uploadBox'>
						{
							imgAddType === 'local' ? <div>
								<label className='uploadArea' htmlFor='localImg' >{imgUrl}</label>
								<input type='file' ref='url' id='localImg' accept={accept} value={urlValue} className='upload'
									onChange={this.handleChange.bind(this, 'imgUrl')} />
							</div>
								: <input type='text' value={urlValue} onChange={this.handleChange.bind(this, 'imgUrl')} placeholder='请输入图片url' className='input' />
						}
						<div className='imgStyleDiv'>
							<label>宽度：</label>
							<input type="text" value={width} onChange={this.handleChange.bind(this, 'width')} className='imgStyleInput' />
						</div>
						<div className='imgStyleDiv' >
							<label>高度：</label>
							<input type="text" value={height} onChange={this.handleChange.bind(this, 'height')} className='imgStyleInput' />
						</div>
					</div>
					<div className='center' >
						<button className='btn btnPrimary' disabled={btnDisabled} onMouseDown={this.handleImgAdd} >确定</button> &nbsp;&nbsp;&nbsp;&nbsp;
						<button className='btn' onMouseDown={this.changeImgVisible} >取消</button>
					</div>
				</div>
			)
		}
		return (
			<div className="FegoEditor-toolbars">
				{
					IMAGE_TYPES.map(type =>
						<ToolBarBtn
							key={type.label} title={type.title}
							onToggle={this.onToggle}
							style={type.style}
						>
							{Iconfont[type.icon]}
						</ToolBarBtn>
					)
				}
				{imageModal}
			</div>
		)
	}
}