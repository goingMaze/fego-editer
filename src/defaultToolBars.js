const defaultToolBars = {
	options: ['BlockType', 'Color', 'FontFamily', 'FontSize', 'Image', 'Inline', 'Link', 'List', 'TextAlign', 'Video', 'History'],
	BgColor: [],
	BlockType: ['正文', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'],
	Color: ['rgb(97,189,109)', 'rgb(26,188,156)', 'rgb(84,172,210)', 'rgb(44,130,201)',
		'rgb(147,101,184)', 'rgb(71,85,119)', 'rgb(204,204,204)', 'rgb(65,168,95)', 'rgb(0,168,133)',
		'rgb(61,142,185)', 'rgb(41,105,176)', 'rgb(85,57,130)', 'rgb(40,50,78)', 'rgb(0,0,0)',
		'rgb(247,218,100)', 'rgb(251,160,38)', 'rgb(235,107,86)', 'rgb(226,80,65)', 'rgb(163,143,132)',
		'rgb(239,239,239)', 'rgb(255,255,255)', 'rgb(250,197,28)', 'rgb(243,121,52)', 'rgb(184,49,47)',
		'rgb(209,72,65)', 'rgb(124,112,107)', 'rgb(209,213,216)'],
	FontFamily: [{
		name: 'SimHei', family: '"SimHei", "黑体"'
	}, {
		name: 'Yahei', family: '"Microsoft Yahei", "微软雅黑"'
	}, {
		name: 'KaiTi', family: '"KaiTi", "楷体"'
	}, {
		name: 'FangSong', family: '"FangSong", "仿宋"'
	}],
	FontSize: [16, 20, 30, 40, 50, 70, 100, 200],
	History: ['undo', 'redo'],
	Image: {
		uploadUrl: '',
		uploadCallback: null
	},
	Inline: ['bold', 'italic', 'underline'],
	Link: {},
	List: ['ul', 'ol', 'indent', 'outdent'],
	TextAlign: ['left', 'center', 'right', 'justify'],
	Video: {},
	Remove: {}
};

export default defaultToolBars
