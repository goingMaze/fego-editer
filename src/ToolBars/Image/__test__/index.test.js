import React from 'react'
import { mount, render } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { EditorState, ContentState, convertFromHTML } from 'draft-js'
import Image from '..'
import defaultToolBars from '../../../default'
import ModalManage from '../../../utils/ModalManage'
import { wrap } from 'module';

describe('Image', () => {
	const contentState = ContentState.createFromBlockArray(convertFromHTML('<div>test</div>'))
	const editorState = EditorState.createWithContent(contentState)
	const wrapper = mount(
		<Image editorState={editorState} onChange={() => { }}
			config={defaultToolBars.Image} modalManage={new ModalManage()}
		/>
	);
	it('渲染时开头应该有一个div', () => {
		expect(wrapper.html().startsWith('<div')).toEqual(true)
	})
	it('应该有1个toolBarBtn子组件', () => {
		expect(wrapper.find('ToolBarBtn').length).toBe(1)
	})
	it('快照', () => {
		const rendered = render(
			<Image editorState={editorState} onChange={() => { }}
				config={defaultToolBars.Image} modalManage={new ModalManage()}
			/>
		)
		expect(toJSON(rendered)).toMatchSnapshot()
	})
})