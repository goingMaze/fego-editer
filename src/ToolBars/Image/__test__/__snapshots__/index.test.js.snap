// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image 快照 1`] = `
<div
  class="FegoEditor-toolbars"
>
  <div
    class="FegoEditor-toolBarBtn "
    title="图片"
  >
    <svg
      aria-hidden="true"
      class="icon"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M901.25224719 124.64246844H122.74583469c-49.64058375 0-89.84692406 40.00967344-89.84692407 89.37204562v595.93547625c0 49.35949406 40.20538031 89.40850125 89.84692407 89.40850031h778.5054525c49.60221 0 89.84692406-40.04900719 89.84692406-89.40850031l0.00191906-595.93547625c-0.03837375-49.36045406-40.27829156-89.37204562-89.84788312-89.37204562z m14.96201531 594.10024031l-187.94617219-249.37332563s-10.90876406-20.17128375-49.27219312-20.17128375c-43.52281125 0-56.60546156 19.54770656-56.60546156 19.54770657L457.76988781 721.60348906s-8.95360969 16.94403188-30.91887656 16.94403188c-22.99656844 0-33.97728375-16.94403188-33.97728375-16.94403188L306.30777969 623.13680937s-22.18303969-29.22754406-51.00286031-29.22754406c-28.70949469 0-52.55220937 32.34447-52.55220938 32.34447094l-95.00438719 117.79565437V230.00591187c0-16.42982062 13.41458531-29.81466562 29.96144625-29.81466562h748.58142188c16.54878 0 29.96144625 13.38388594 29.96144531 29.81466563v488.73775593h-0.03837375v-0.00095906z m-449.22886219-325.91023313c0 58.30926656-47.46573844 105.54284156-106.02347625 105.54284157s-106.06280906-47.27003062-106.06281-105.54284157c0-58.23827438 47.50411219-105.51022406 106.06281-105.51022406 58.55773781 0.04029281 106.02347625 47.27386875 106.02347625 105.51022406z"
      />
    </svg>
  </div>
</div>
`;

exports[`Image 快照 1`] = `
<div
  class="FegoEditor-toolbars"
>
  <div
    class="FegoEditor-toolBarBtn "
    title="图片"
  >
    <svg
      aria-hidden="true"
      class="icon"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M901.25224719 124.64246844H122.74583469c-49.64058375 0-89.84692406 40.00967344-89.84692407 89.37204562v595.93547625c0 49.35949406 40.20538031 89.40850125 89.84692407 89.40850031h778.5054525c49.60221 0 89.84692406-40.04900719 89.84692406-89.40850031l0.00191906-595.93547625c-0.03837375-49.36045406-40.27829156-89.37204562-89.84788312-89.37204562z m14.96201531 594.10024031l-187.94617219-249.37332563s-10.90876406-20.17128375-49.27219312-20.17128375c-43.52281125 0-56.60546156 19.54770656-56.60546156 19.54770657L457.76988781 721.60348906s-8.95360969 16.94403188-30.91887656 16.94403188c-22.99656844 0-33.97728375-16.94403188-33.97728375-16.94403188L306.30777969 623.13680937s-22.18303969-29.22754406-51.00286031-29.22754406c-28.70949469 0-52.55220937 32.34447-52.55220938 32.34447094l-95.00438719 117.79565437V230.00591187c0-16.42982062 13.41458531-29.81466562 29.96144625-29.81466562h748.58142188c16.54878 0 29.96144625 13.38388594 29.96144531 29.81466563v488.73775593h-0.03837375v-0.00095906z m-449.22886219-325.91023313c0 58.30926656-47.46573844 105.54284156-106.02347625 105.54284157s-106.06280906-47.27003062-106.06281-105.54284157c0-58.23827438 47.50411219-105.51022406 106.06281-105.51022406 58.55773781 0.04029281 106.02347625 47.27386875 106.02347625 105.51022406z"
      />
    </svg>
  </div>
</div>
`;
